File: windows_app/src/core/discovery.py
"""
UDP Discovery Service
Allows clients to find the server on local network
"""
import asyncio
import socket
import json
from utils.logger import get_logger

logger = get_logger(__name__)

class DiscoveryService:
    """UDP broadcast service for device discovery"""
    
    DISCOVERY_PORT = 37020
    MAGIC_BYTES = b'NEXREMOTE_DISCOVER'
    
    def __init__(self, config):
        self.config = config
        self.running = False
        self.socket = None
        
    async def start(self, response_callback):
        """Start listening for discovery requests"""
        try:
            self.running = True
            
            # Create UDP socket
            self.socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
            self.socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
            self.socket.bind(('', self.DISCOVERY_PORT))
            
            logger.info(f"Discovery service listening on port {self.DISCOVERY_PORT}")
            
            while self.running:
                try:
                    # Receive discovery request
                    data, addr = await asyncio.get_event_loop().run_in_executor(
                        None, self.socket.recvfrom, 1024
                    )
                    
                    if data.startswith(self.MAGIC_BYTES):
                        logger.info(f"Discovery request from {addr}")
                        
                        # Get response from callback
                        response = response_callback(addr)
                        
                        # Send response
                        response_data = json.dumps(response).encode('utf-8')
                        await asyncio.get_event_loop().run_in_executor(
                            None, self.socket.sendto, response_data, addr
                        )
                        
                        logger.debug(f"Sent discovery response to {addr}")
                        
                except Exception as e:
                    if self.running:
                        logger.error(f"Error handling discovery request: {e}")
                    
        except Exception as e:
            logger.error(f"Discovery service error: {e}", exc_info=True)
        finally:
            if self.socket:
                self.socket.close()
    
    def stop(self):
        """Stop the discovery service"""
        self.running = False
        logger.info("Discovery service stopped")

File: mobile_app/lib/core/discovery.dart
import 'dart:io';
import 'dart:async';
import 'dart:convert';
import '../utils/logger.dart';

class DiscoveryService {
  static const int discoveryPort = 37020;
  static const String magicBytes = 'NEXREMOTE_DISCOVER';
  
  Future<List<ServerInfo>> discoverServers({Duration timeout = const Duration(seconds: 5)}) async {
    final servers = <ServerInfo>[];
    
    try {
      // Create UDP socket
      final socket = await RawDatagramSocket.bind(InternetAddress.anyIPv4, 0);
      socket.broadcastEnabled = true;
      
      // Send broadcast
      final request = utf8.encode(magicBytes);
      socket.send(request, InternetAddress('255.255.255.255'), discoveryPort);
      
      Logger.info('Sent discovery broadcast');
      
      // Listen for responses
      final completer = Completer<List<ServerInfo>>();
      
      socket.listen((RawSocketEvent event) {
        if (event == RawSocketEvent.read) {
          final datagram = socket.receive();
          if (datagram != null) {
            try {
              final response = utf8.decode(datagram.data);
              final data = jsonDecode(response);
                            if (data['type'] == 'discovery_response') {
                final server = ServerInfo(
                  name: data['name'] ?? 'Unknown PC',
                  address: datagram.address.address,
                  port: data['port'] ?? 8765,
                  portInsecure: data['port_insecure'] ?? 8766,
                  id: data['id'] ?? '',
                  version: data['version'] ?? '1.0.0',
                );
                
                // Avoid duplicates
                if (!servers.any((s) => s.id == server.id)) {
                  servers.add(server);
                  Logger.info('Found server: ${server.name} at ${server.address}');
                }
              }
            } catch (e) {
              Logger.error('Error parsing discovery response: $e');
            }
          }
        }
      });
      
      // Timeout
      Future.delayed(timeout, () {
        if (!completer.isCompleted) {
          socket.close();
          completer.complete(servers);
        }
      });
      
      return completer.future;
      
    } catch (e) {
      Logger.error('Discovery error: $e');
      return [];
    }
  }
}

class ServerInfo {
  final String name;
  final String address;
  final int port;
  final int portInsecure;
  final String id;
  final String version;
  
  ServerInfo({
    required this.name,
    required this.address,
    required this.port,
    required this.portInsecure,
    required this.id,
    required this.version,
  });
}